src/navigation/water_bottle_search.py

```diff
@@ -0,0 +1,426 @@
+
+            outer_iterations += 1
+
+            if outer_iterations > max_iterations:
+                # if we hit this point return the path such as it is. It will not contain the destination
+                print("giving up on pathfinding too many iterations")
+                return self.return_path(current_node, costmap2d)
+
+            # get the current node
+            current_node = heapq.heappop(open_list)
+            closed_list.append(current_node)
+
+            # found the goal
+            if current_node == end_node:
+                return self.return_path(current_node, costmap2d)
+
+            # generate children
+            children = []
+            for pick_index in adjacent_square_pick_index:
+                new_position = adjacent_squares[pick_index]
+                node_position = (current_node.position[0] + new_position[0], current_node.position[1] + new_position[1])
+                # make sure within range
+                if (
+                    node_position[0] > (costmap2d.shape[0] - 1)
+                    or node_position[0] < 0
+                    or node_position[1] > (costmap2d.shape[1] - 1)
+                    or node_position[1] < 0
+                ):
+                    continue
+
+                # make sure it is traversable terrian (not too high of a cost)
+                if costmap2d[node_position[0]][node_position[1]] >= 0.2: # TODO: find optimal value
+                    continue
+
+                # create new node and append it
+                new_node = Node(current_node, node_position)
+                children.append(new_node)
+
+            # loop through children
+            for child in children:
+                # child is on the closed list
+                if len([closed_child for closed_child in closed_list if closed_child == child]) > 0:
+                    continue
+                # create the f (total), g (cost in map), and h (euclidean distance) values
+                child.g = current_node.g + costmap2d[child.position[0], child.position[1]]
+                child.h = ((child.position[0] - end_node.position[0]) ** 2) + (
+                    (child.position[1] - end_node.position[1]) ** 2
+                )
+                child.f = child.g + child.h
+                # child is already in the open list
+                if (
+                    len(
+                        [
+                            open_node
+                            for open_node in open_list
+                            if child.position == open_node.position and child.g > open_node.g
+                        ]
+                    )
+                    > 0
+                ):
+                    continue
+                # add the child to the open list
+                heapq.heappush(open_list, child)
+
+        print("Couldn't find a path to destination")
+        raise NoPath()
+
+    def costmap_callback(self, msg: OccupancyGrid):
```

Comment:
the costmap should be stored in the context, so you should have this callback their too. Then you can pass the context.costmap into the AStar object to get the path